lib/contracts/provider.js

```diff
@@ -1,98 +1,67 @@
-          });
-          return;
-        }
-        this.accounts.push(account);
-        this.addresses.push(account.address);
-      });
-
-      if (this.accounts.length) {
-        this.accounts.forEach(account => {
-          this.web3.eth.accounts.wallet.add(account);
-        });
-        this.asyncMethods = {
-            eth_accounts: self.eth_accounts.bind(this)
-        };
-        if (this.isDev) {
-          this.logger.warn('You are using your own account in the develop environment. It might not be funded.');
-        }
-      }
-    }
-
     // network connectivity error
-    this.engine.on('error', (err) => {
+    self.engine.on('error', (err) => {
       // report connectivity errors
-      this.logger.error(err.stack);
+      self.logger.error(err.stack);
     });
-    this.engine.start();
-  }
-
-  getAccount(accountConfig) {
-    if (accountConfig.privateKey) {
-      if (!accountConfig.privateKey.startsWith('0x')) {
-        accountConfig.privateKey = '0x' + accountConfig.privateKey;
-      }
-      return this.web3.eth.accounts.privateKeyToAccount(accountConfig.privateKey);
-    }
-    if (accountConfig.privateKeyFile) {
-      let fileContent = fs.readFileSync(fs.dappPath(accountConfig.privateKeyFile)).toString();
-      fileContent = fileContent.trim().split(/[,;]/);
-      return fileContent.map(key => {
-        if (!key.startsWith('0x')) {
-          key = '0x' + key;
-        }
-        return this.web3.eth.accounts.privateKeyToAccount(key);
-      });
-    }
-    if (accountConfig.mnemonic) {
-      const hdwallet = hdkey.fromMasterSeed(bip39.mnemonicToSeed(accountConfig.mnemonic.trim()));
 
-      const addressIndex = accountConfig.addressIndex || 0;
-      const numAddresses = accountConfig.numAddresses || 1;
-      const wallet_hdpath = accountConfig.hdpath || "m/44'/60'/0'/0/";
+    self.engine.start();
+    self.web3.setProvider(self);
 
-      const accounts = [];
-      for (let i = addressIndex; i < addressIndex + numAddresses; i++) {
-        const wallet = hdwallet.derivePath(wallet_hdpath + i).getWallet();
-        accounts.push(this.web3.eth.accounts.privateKeyToAccount('0x' + wallet.getPrivateKey().toString('hex')));
+    self.accounts = AccountParser.parseAccountsConfig(self.accountsConfig, self.web3, self.logger);
+    self.addresses = [];
+    async.waterfall([
+      function fundAccounts(next) {
+        if (!self.accounts.length) {
+          return next(NO_ACCOUNTS);
+        }
+        if (!self.isDev) {
+          return next();
+        }
+        async.each(self.accounts, (account, eachCb) => {
+          fundAccount(self.web3, account.address, eachCb);
+        }, next);
+      },
+      function populateWeb3Wallet(next) {
+        self.accounts.forEach(account => {
+          self.addresses.push(account.address);
+          self.web3.eth.accounts.wallet.add(account);
+        });
+        self.asyncMethods = {
+          eth_accounts: self.eth_accounts.bind(self)
```

Comment:
why is the bind method used here?

Reply:
To make sure `eth_accounts` uses the right this, because we need to have access to `this.accounts`

Reply:
okay.